client/asset/zec/zec.go

```diff
@@ -92,238 +114,2323 @@ var (
+		ops[i] = btc.NewOutput(&txHash, uint32(i), uint64(tx.TxOut[i].Value))
+		locks[i] = &btc.UTxO{
+			TxHash:  &txHash,
+			Vout:    uint32(i),
+			Amount:  uint64(tx.TxOut[i].Value),
+			Address: outputAddresses[i].String(),
+		}
+	}
+	w.cm.LockOutputs(locks)
+	lockUnspent(w, false, ops)
+
+	var totalOut uint64
+	for _, txOut := range tx.TxOut {
+		totalOut += uint64(txOut.Value)
+	}
+
+	success = true
+	return coins, totalIn - totalOut, nil
+}
+
+func (w *zecWallet) fundsRequiredForMultiOrders(orders []*asset.MultiOrderValue, inputCount, inputsSize uint64) ([]uint64, uint64) {
+	requiredForOrders := make([]uint64, len(orders))
+	var totalRequired uint64
+
+	for i, value := range orders {
+		req := dexzec.RequiredOrderFunds(value.Value, inputCount, inputsSize, value.MaxSwapCount)
+		requiredForOrders[i] = req
+		totalRequired += req
+	}
+
+	return requiredForOrders, totalRequired
+}
+
+// fundMultiSplitTx uses the utxos provided and attempts to fund a multi-split
+// transaction to fund each of the orders. If successful, it returns the
+// funding coins and outputs.
+func (w *zecWallet) fundMultiSplitTx(
+	orders []*asset.MultiOrderValue,
+	utxos []*btc.CompositeUTXO,
+	keep, maxLock uint64,
+) (bool, asset.Coins, []*btc.Output) {
+
+	_, totalOutputRequired := w.fundsRequiredForMultiOrders(orders, uint64(len(utxos)), dexbtc.RedeemP2PKHInputSize)
+
+	outputsSize := uint64(dexbtc.P2WPKHOutputSize) * uint64(len(utxos)+1)
+	// splitTxSizeWithoutInputs := dexbtc.MinimumTxOverhead + outputsSize
+
+	enough := func(inputCount, inputsSize, sum uint64) (bool, uint64) {
+		fees := dexzec.TxFeesZIP317(inputsSize+uint64(wire.VarIntSerializeSize(inputCount)), outputsSize, 0, 0, 0, 0)
+		req := totalOutputRequired + fees
+		return sum >= req, sum - req
+	}
+
+	fundSplitCoins, _, spents, _, inputsSize, _, err := w.cm.FundWithUTXOs(utxos, keep, false, enough)
+	if err != nil {
+		return false, nil, nil
+	}
+
+	if maxLock > 0 {
+		fees := dexzec.TxFeesZIP317(inputsSize+uint64(wire.VarIntSerializeSize(uint64(len(spents)))), outputsSize, 0, 0, 0, 0)
+		if totalOutputRequired+fees > maxLock {
+			return false, nil, nil
+		}
+	}
+
+	return true, fundSplitCoins, spents
+}
+
+func (w *zecWallet) Info() *asset.WalletInfo {
+	return WalletInfo
+}
+
+func (w *zecWallet) LockTimeExpired(_ context.Context, lockTime time.Time) (bool, error) {
+	chainStamper := func(blockHash *chainhash.Hash) (stamp time.Time, prevHash *chainhash.Hash, err error) {
+		hdr, err := getBlockHeader(w, blockHash)
+		if err != nil {
+			return
+		}
+		return hdr.Timestamp, &hdr.PrevBlock, nil
+	}
+
+	w.tipMtx.RLock()
+	tip := w.currentTip
+	w.tipMtx.RUnlock()
+
+	medianTime, err := btc.CalcMedianTime(chainStamper, &tip.Hash) // TODO: pass ctx
+	if err != nil {
+		return false, fmt.Errorf("error getting median time: %w", err)
+	}
+	return medianTime.After(lockTime), nil
+}
+
+// fundMultiOptions are the possible order options when calling FundMultiOrder.
+type fundMultiOptions struct {
```

Comment:
Why doesn't zcash have a fund multi buffer?

Reply:
The buffer is for dealing with varying fee rates. Zcash doesn't have fee rates anymore. That said, it's possible that a buffer of "logical actions" is appropriate, in case more shielded notes are later required. Leaving for now.